Feature - Optimized N+1 problems and added Integration tests

wickathou opened the pull request:
# Features
- [x] Fixed N+1 query issues on multiple views
- [x] Added new integration tests for all the views in the app

# N+1 issues found and fixed
- [x] Comments belonging to a post triggered multiple queries to find the user
![image](https://github.com/wickathou/mcv-blog/assets/50066876/db348efb-e1cc-4dca-9f55-28b0dc2a7e0d)
- [x] Posts author triggers a N+1 query
![image](https://github.com/wickathou/mcv-blog/assets/50066876/be3bd29e-45b6-4713-8f42-6ae6ddadbe70)

anshuman7negi committed:
create branch

wickathou committed:
Updated partials for user_listing to include profile picture
Added pagination button for post index page
Added author name to post show page
Added post and user index and show integration tests with capybara

wickathou committed:
Merge pull request #8 from wickathou/integration-tests

Inner merge

anshuman7negi committed:
solved n+1 problem

Ibizugbe reviewed (commented):
Hi @Team👋🏼👋🏼, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

It seems you have made a mistake submitting the wrong link or submitting your project in the wrong place. this is milestone on integration test and has been submitted in the `add devise` milestone. kindly do well to submit the right link.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

tresorsawasawa commented on app/controllers/posts_controller.rb:
```diff
@@ -1,15 +1,15 @@
 class PostsController < ApplicationController
   def index
     @user = User.find(params[:user_id])
-    @posts = Post.where(author: @user)
+    @posts = Post.includes(:author).where(author: @user)
     respond_to do |format|
       format.html { render 'posts/index' }
       format.json { render json: { user: @user, posts: @posts } }
     end
   end
 
   def show
-    @post = Post.find(params[:id])
+    @post = Post.includes(comments: [:user]).find(params[:id])
```
- [x] Good job by fixing the N+1 problem when fetching all posts and their comments without using the `bullet gem. However as required for the project, kindly ensure that your PR description includes the screenshots of your app's logs **before** and **after** the fix

tresorsawasawa commented on spec/system/user_integrations_spec.rb:
```diff
@@ -0,0 +1,71 @@
+require 'rails_helper'
+
+RSpec.describe 'user show page tests', type: :system do
+  test_user = User.create!(name: 'Test User', bio: 'With a bio')
+  post_one = Post.create!(title: 'Test Post 1', text: 'With some text 1', author: test_user)
+  post_two = Post.create!(title: 'Test Post 2', text: 'With some text 2', author: test_user)
+  post_three = Post.create!(title: 'Test Post 3', text: 'With some text 3', author: test_user)
+  post_four = Post.create!(title: 'Test Post 4', text: 'With some text 4', author: test_user)
+  post_five = Post.create!(title: 'Test Post 5', text: 'With some text 5', author: test_user)
+
+  describe 'index page' do
+    it 'shows the right content' do
+      visit users_path
+      expect(page).to have_content('Users')
+    end
+  end
```
- [x] As required for the project, kindly ensure that you have a Spec file which only tests `User` index page.

- [x] Once you have create that file, you should ensure that, it includes all the following tests:
  * Tests that asserts you can see the username of all other users.
  * Tests that asserts you can see the profile picture for each user.
  * Tests that asserts you can see the number of posts each user has written.
  * Tests that redirects to the user's show page when you click on a user.

tresorsawasawa commented on spec/system/user_integrations_spec.rb:
```diff
@@ -0,0 +1,71 @@
+      visit user_path(test_user)
+      expect(page).to have_css('div.profile-picture')
+    end
+
+    it 'check that user page shows the user bio' do
+      visit user_path(test_user)
+      expect(page).to have_content(test_user.bio)
+    end
+
+    it 'check that user page shows the last 3 posts' do
+      visit user_path(test_user)
+      expect(page).to have_content(post_three.title)
+      expect(page).to have_content(post_four.title)
+      expect(page).to have_content(post_five.title)
+      expect(page).not_to have_content(post_one.title)
+      expect(page).not_to have_content(post_two.title)
+    end
+
+    it 'check if button exist in the user page with the string See all posts' do
+      visit user_path(test_user)
+      expect(page).to have_link('See all posts')
+    end
+
+    it 'checks that clicking on a post title takes you to the post page' do
+      visit user_path(test_user)
+      click_on(post_three.title)
+      expect(page).to have_content(post_three.title)
+      expect(page).to have_content(post_three.text)
+    end
+
+    it 'check that user can click on button to view all posts' do
+      visit user_path(test_user)
+      click_on('See all posts')
+      expect(page).to have_content(post_one.title)
+      expect(page).to have_content(post_two.title)
+      expect(page).to have_content(post_three.title)
+      expect(page).to have_content(post_four.title)
+      expect(page).to have_content(post_five.title)
+    end
+  end
```
- [x] As required for the project, kindly ensure that you have a Spec file which only tests `User` Show page.

tresorsawasawa commented on spec/system/post_integrations_spec.rb:
```diff
@@ -0,0 +1,118 @@
+
+    it 'check that a post title is rendered' do
+      visit user_posts_path(test_user_one)
+      expect(page).to have_content(post_user_one.title)
+    end
+
+    it 'check that a post text is rendered' do
+      visit user_posts_path(test_user_one)
+      expect(page).to have_content(post_user_one.text)
+    end
+
+    it 'check that comments are rendered' do
+      visit user_posts_path(test_user_one)
+      expect(page).to have_content(comment_user_one.text)
+    end
+
+    it 'checks that the comment count is shown' do
+      visit user_posts_path(test_user_one)
+      expect(page).to have_content('Comments:')
+      expect(page).to have_content(post_user_one.comments_count)
+    end
+
+    it 'checks that the likes count is shown' do
+      visit user_posts_path(test_user_one)
+      expect(page).to have_content('Likes:')
+      expect(page).to have_content(post_user_one.likes_count)
+    end
+
+    it 'checks that the pagination button is shown' do
+      visit user_posts_path(test_user_one)
+      expect(page).to have_button('Load more posts')
+    end
+
+    it 'checks that when clicking on a post title takes you to the post page' do
+      visit user_posts_path(test_user_one)
+      click_on(post_user_one.title)
+      expect(page).to have_content(post_user_one.title)
+      expect(page).to have_content(post_user_one.text)
+    end
+  end
```
- [x] As required for the project, kindly ensure that you have a Spec file which only tests `Post` index page.

tresorsawasawa commented on spec/system/post_integrations_spec.rb:
```diff
@@ -0,0 +1,118 @@
+    it 'shows the title of the post' do
+      visit user_post_path(test_user_one, post_user_one)
+      expect(page).to have_content(post_user_one.title)
+    end
+
+    it 'shows the username of the author' do
+      visit user_post_path(test_user_one, post_user_one)
+      expect(page).to have_content(test_user_one.name)
+    end
+
+    it 'checks that the comment count is shown' do
+      visit user_post_path(test_user_one, post_user_one)
+      expect(page).to have_content('Comments:')
+      expect(page).to have_content(post_user_one.comments_count)
+    end
+
+    it 'checks that the likes count is shown' do
+      visit user_post_path(test_user_one, post_user_one)
+      expect(page).to have_content('Likes:')
+      expect(page).to have_content(post_user_one.likes_count)
+    end
+
+    it 'shows the body text of the post' do
+      visit user_post_path(test_user_one, post_user_one)
+      expect(page).to have_content(post_user_one.text)
+    end
+
+    it 'shows the names of the people leaving comments on the post' do
+      visit user_post_path(test_user_one, post_user_one)
+      within('ul.comments') do
+        expect(page).to have_content(test_user_two.name)
+      end
+    end
+
+    it 'shows the text of the comments on the post' do
+      visit user_post_path(test_user_one, post_user_one)
+      within('ul.comments') do
+        expect(page).to have_content(comment_user_one.text)
+      end
+    end
```
- [x] As required for the project, kindly ensure that you have a Spec file which only tests `User` ** post show page**.

tresorsawasawa commented on Gemfile.lock:
```diff
@@ -257,6 +262,7 @@ GEM
 PLATFORMS
+  x64-mingw-ucrt
   x86_64-linux
 
 DEPENDENCIES
```
**README.md** file:

- [x] To make your README looks more professional and as required the used template, kindly ensure that you update the `## Authors` section of your README by adding all authors contributors of the project

tresorsawasawa reviewed (changes requested):
# STATUS: REQUIRED CHANGES :recycle::recycle: 

Hi @wickathou  and @anshuman7negi  
Good job so far! I know you worked hard to implement this project, I really appreciate the effort you put in! But there are some issues that you still need to work on to go to the next project but you are almost there! 

![almost-there](https://user-images.githubusercontent.com/57562869/143511358-9cde7daf-4ad0-4db2-8f15-18b54288ff91.gif)

## To highlight :champagne: 
- All tests are passing :heavy_check_mark: 
- Most of required asserts are present :heavy_check_mark: 
- Follow GitHub flow :heavy_check_mark: 

## Required changes :recycle: :recycle: 
- [x] Check the comments under the review.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

wickathou committed:
Added bullet gem and saved screenshots showing the N+1 issues
Added collaborator to readme file
Separated tests for Posts and Users based on the view tested

wickathou committed:
fixed lint issues

codepantha commented on spec/system/user_show_integrations_spec.rb:
```diff
@@ -0,0 +1,64 @@
+      expect(page).to have_content(test_user.bio)
+    end
+
+    it 'check that user page shows the last 3 posts' do
+      visit user_path(test_user)
+      expect(page).to have_content(post_three.title)
+      expect(page).to have_content(post_four.title)
+      expect(page).to have_content(post_five.title)
+      expect(page).not_to have_content(post_one.title)
+      expect(page).not_to have_content(post_two.title)
+    end
```
- [x] Well done for creating a `spec` file which only tests the user show page. However, for this particular test,  you need to write tests that asserts that you can see the user's first 3 posts not their last 3 posts as you have done here 👍 
![image](https://github.com/wickathou/mcv-blog/assets/46839548/cd6cf799-2c93-4ecf-898f-1a42aca8f68c)

codepantha commented on spec/system/post_show_integrations_spec.rb:
```diff
@@ -0,0 +1,61 @@
+      expect(page).to have_content(post_user_one.text)
+    end
+
+    it 'shows the names of the people leaving comments on the post' do
+      visit user_post_path(test_user_one, post_user_one)
+      within('ul.comments') do
+        expect(page).to have_content(test_user_two.name)
+      end
+    end
+
+    it 'shows the text of the comments on the post' do
+      visit user_post_path(test_user_one, post_user_one)
+      within('ul.comments') do
+        expect(page).to have_content(comment_user_one.text)
+      end
+    end
```
- [x] For these two tests, in order for us to truly test that the assertions are true, we need to ensure that we have at least two comments from different users. Only then can we have the rest of mind and assurance that comes with our tests passing.
![image](https://github.com/wickathou/mcv-blog/assets/46839548/10388411-322a-412f-8c97-0221e61d70bc)

codepantha reviewed (changes requested):
# STATUS: CHANGES REQUIRED ♻️ 

## Hi team 👋, 

Great work on making the changes requested by the previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

<img src="https://media4.giphy.com/media/rZ10JtoPRv1Ly/giphy.gif"/>

## Highlights 🎯 
✔️ Descriptive PR
✔️ All tests are passing 🟢
✔️ No linter errors

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

wickathou committed:
Added new method to return the first 3 posts of an user
Modified Post_show test to check if all commenters username and text are shown

wickathou committed:
Fixed test for user show

Ibizugbe reviewed (changes requested):
Hi @Team👋🏼👋🏼👋🏼, 

![request change](https://github.com/wickathou/mcv-blog/assets/64643995/bb5593a0-e76b-4b34-8149-5a1b3d0c8122)


Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights✅✅✅

- [x] The test that asserts we can see the first 3 posts has been updated👍🏼.
- [x] Linters are still passing👍🏼.


 ## Required Changes ♻️

- [ ] Nice job so far😉. Some of your test cases are still failing, kindly ensure all tests are passing before proceeding to make submissions to avoid blockers as this is a requirement that all tests should be passing😉.
<img width="751" alt="image" src="https://github.com/wickathou/mcv-blog/assets/64643995/ed25ba66-695c-4687-b8d5-6b5bf41b8717">


_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

wickathou committed:
Modified queries to avoid n+1 queries
Fixed issues found with the integrations tests

FrederickMih reviewed (approved):
## Hi :raising_hand: @wickathou, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🎉 

:tada: :champagne: <img src="https://media.giphy.com/media/LtkRsMZNSChvICUFhV/giphy.gif"  width ="160"><img> :tada: :champagne: 

## Highlights :green_circle: 

- Tests covered all the user workflow :+1: 

## Optional suggestions
**N/A**

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
